File: python/tests/connection/smbus_test.py
import sys

from periph.connection.smbus_auto import SMBusConnection

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


# --- offline: address validation ---
# SMBus restricts addresses to 0x08–0x77; addresses outside this range must
# be rejected before any bus access.

try:
    SMBusConnection(0x07)
    check_true('addr 0x07 rejected', False)
except ValueError:
    check_true('addr 0x07 rejected', True)

try:
    SMBusConnection(0x78)
    check_true('addr 0x78 rejected', False)
except ValueError:
    check_true('addr 0x78 rejected', True)

# --- online: bus scan ---

found = []
for probe_addr in range(0x08, 0x78):
    t = SMBusConnection(probe_addr)
    try:
        t.read(1)
        found.append(probe_addr)
    except OSError:
        pass
    finally:
        t.close()

check_true('bus scan: at least one device found', len(found) > 0)
print('devices found:', [hex(a) for a in found])

# --- online: PEC ---
# PEC appends a CRC-8 byte to every write. Whether the device under test
# accepts the extra byte is not a connection concern; we verify the connection
# computes and sends the CRC without crashing.

if found:
    t = SMBusConnection(found[0], pec=True)
    try:
        t.write(bytes([0x00]))
    except OSError:
        pass  # device rejected CRC-appended byte; connection layer worked correctly
    finally:
        t.close()
    check_true('PEC write: no connection crash', True)

print('===DONE: {} passed, {} failed==='.format(passed, failed))
try:
    sys.exit(0 if failed == 0 else 1)
except AttributeError:
    pass  # MicroPython has no sys.exit

File: nodejs/tests/connection/smbus_test.js
'use strict';

const { SMBusConnection } = require('../../packages/periph/src/connection/smbus');

const I2C_BUS  = parseInt(process.env.I2C_BUS  || '1',  10);
const I2C_ADDR = parseInt(process.env.I2C_ADDR  || '0x40', 16);

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else           { console.log('FAIL', label); failed++; }
}

async function main() {
    // --- offline: address validation ---

    try {
        new SMBusConnection(I2C_BUS, 0x07);
        checkTrue('addr 0x07 rejected', false);
    } catch (e) {
        checkTrue('addr 0x07 rejected', e instanceof RangeError);
    }

    try {
        new SMBusConnection(I2C_BUS, 0x78);
        checkTrue('addr 0x78 rejected', false);
    } catch (e) {
        checkTrue('addr 0x78 rejected', e instanceof RangeError);
    }

    // --- online: basic I/O without PEC ---

    const connection = new SMBusConnection(I2C_BUS, I2C_ADDR);

    const readBuf = await connection.read(1);
    checkTrue('read returns 1 byte', readBuf.length === 1);

    await connection.write(Buffer.from([0x00]));
    checkTrue('write accepted', true);

    const wrBuf = await connection.writeRead(Buffer.from([0x00]), 1);
    checkTrue('writeRead returns 1 byte', wrBuf.length === 1);

    // --- online: write with PEC enabled ---

    const connectionPec = new SMBusConnection(I2C_BUS, I2C_ADDR, true);
    await connectionPec.write(Buffer.from([0x00]));
    checkTrue('write with PEC accepted', true);
    await connectionPec.close();

    await connection.close();

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();
